driver/registry_memory.go

```diff
@@ -102,6 +106,41 @@ func (r *RegistryMemory) RuleMatcher() rule.Matcher {
 	return r.ruleRepository
 }
 
+func (r *RegistryMemory) UpstreamTransport(req *http.Request) (http.RoundTripper, error) {
+
+	// Use req to decide the transport per request iff need be.
+
+	certFile := r.c.ProxyServeUpstreamCaAppendCrtPath()
```

Comment:
It looks like this would be called on every `UpstreamTransport` which means reading the cert file on every single request. This will cause significant issues at scale. Instead, we should cache the TLS certification / transport. Keep in mind that the config may change at any point in time so it is probably a good idea to invalidate the cache if the `certFile` value has changed.

Reply:
@wraix you can put the cached transport in a var that belongs to `RegistryMemory` so something like:

```
type RegistryMemory {
  upstreamTransport *http.Transport
  cachedCertFilePath string
}
```

and use `cachedCertFilePath` to check if the file location changed so that you can invalidate the cache. Alternatively you could also `os.Stat` the file to check if its mod time was changed, but this would again require disk access which is expensive. As an alternative, you could make the `os.Stat` check every 1 / 1000 requests (or something similar) giving a decent trade off between cache staleness and speed.

Reply:
Might do both and put some configuration for the os.Stat check value.

Thanks for the pointer - will give it a Go.

Reply:
@aeneasr  I made an implementation. Not sure if test cases for the certificate refresh is needed or how to perform those in a test suite.